File: TerraObserver/Scenes/Geos/Views/LonLatGrid.cs
using System;
using Godot;
using TO.Abstractions.Views.Geos;
using TO.Presenters.Views.Geos;

namespace TerraObserver.Scenes.Geos.Views;

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-04 10:48:55
[Tool]
public partial class LonLatGrid : Node3D, ILonLatGrid
{
    #region 事件

    public event Action<bool>? FixFullVisibilityChanged;
    public event Action? DoDrawRequested;

    #endregion

    #region Export 属性

    [ExportToolButton("手动触发重绘经纬线", Icon = "WorldEnvironment")]
    public Callable Redraw => Callable.From(() => DoDrawRequested?.Invoke());

    [Export(PropertyHint.Range, "0, 180")] public int LongitudeInterval { get; set; } = 15;
    [Export(PropertyHint.Range, "0, 90")] public int LatitudeInterval { get; set; } = 15;
    [Export(PropertyHint.Range, "1, 100")] public int Segments { get; set; } = 30; // 每个 90 度的弧线被划分多少段
    [Export] public Material? LineMaterial { get; set; }

    [ExportGroup("颜色设置")] [Export] public Color NormalLineColor { get; set; } = Colors.SkyBlue with { A8 = 40 };

    [Export] public Color DeeperLineColor { get; set; } = Colors.DeepSkyBlue with { A8 = 40 };
    [Export] public int DeeperLineInterval { get; set; } = 3; // 更深颜色的线多少条出现一次
    [Export] public Color TropicColor { get; set; } = Colors.Green with { A8 = 40 }; // 南北回归线颜色
    [Export] public Color CircleColor { get; set; } = Colors.Aqua with { A8 = 40 }; // 南北极圈颜色
    [Export] public Color EquatorColor { get; set; } = Colors.Yellow with { A8 = 40 }; // 赤道颜色
    [Export] public Color Degree90LongitudeColor { get; set; } = Colors.Orange with { A8 = 40 }; // 东西经 90 度颜色
    [Export] public Color MeridianColor { get; set; } = Colors.Red with { A8 = 40 }; // 子午线颜色

    [ExportGroup("开关特定线显示")] [Export] public bool DrawTropicOfCancer { get; set; } = true; // 是否绘制北回归线
    [Export] public bool DrawTropicOfCapricorn { get; set; } = true; // 是否绘制南回归线
    [Export] public bool DrawArcticCircle { get; set; } = true; // 是否绘制北极圈
    [Export] public bool DrawAntarcticCircle { get; set; } = true; // 是否绘制南极圈

    [ExportGroup("透明度设置")]
    [Export(PropertyHint.Range, "0.01, 5")]
    public float FullVisibilityTime { get; set; } = 1.2f;

    [Export(PropertyHint.Range, "0, 1")] public float FullVisibility { get; set; } = 0.6f;
    private bool _fixFullVisibility;

    // 是否锁定开启完全显示
    [Export]
    public bool FixFullVisibility
    {
        get => _fixFullVisibility;
        set
        {
            _fixFullVisibility = value;
            if (value)
            {
                Visibility = FullVisibility;
                FadeVisibility = false;
                Show();
            }
            else
                SetProcess(true);

            if (_ready)
                FixFullVisibilityChanged?.Invoke(value);
        }
    }

    #endregion

    #region 普通属性

    // 对应着色器中的 alpha_factor
    private float _visibility = 1f;

    public float Visibility
    {
        get => _visibility;
        set
        {
            _visibility = Mathf.Clamp(value, 0f, FullVisibility);
            if (_ready && LineMaterial is ShaderMaterial shaderMaterial)
                shaderMaterial.SetShaderParameter("alpha_factor", _visibility);
        }
    }

    public bool FadeVisibility { get; set; }

    public float Radius { get; set; } = 110f;
    public MeshInstance3D? MeshIns { get; set; }

    #endregion

    #region 生命周期

    private bool _ready;

    public override void _Ready()
    {
        MeshIns = new MeshInstance3D();
        AddChild(MeshIns);
        _ready = true;
        // 在 _ready = true 后面，触发 setter 的着色器参数初始化
        Visibility = FullVisibility;
    }

    public override void _Process(double delta)
    {
        if (Engine.IsEditorHint() || FixFullVisibility)
        {
            SetProcess(false); // 编辑器下以及锁定显示时，无需处理显隐
            return;
        }

        if (FadeVisibility)
            Visibility -= (float)delta / FullVisibilityTime;
        FadeVisibility = true;

        if (Visibility > 0) return;
        Hide();
        SetProcess(false);
    }

    #endregion
}

File: TerraObserver/TO.Domains/Functions/Geos/LonLatGrid.fs
namespace TO.Domains.Functions.Geos

open Godot
open TO.Domains.Types.Configs
open TO.Domains.Types.Geos

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-24 22:28:24
module LonLatGridCommand =
    let private draw90Degrees
        (env: #ILonLatGridQuery)
        (surfaceTool: SurfaceTool)
        (color: Color)
        (fromV: Vector3)
        (toV: Vector3)
        (origin: Vector3)
        (dash: bool)
        =
        let lonLatGrid = env.LonLatGrid
        let mutable preDirection = fromV

        for i in 1 .. lonLatGrid.Segments do
            let currentDirection = fromV.Slerp(toV, float32 i / float32 lonLatGrid.Segments)

            if not dash || i % 2 = 0 then
                surfaceTool.SetColor color
                // 【切记】：Mesh.PrimitiveType.Lines 绘制方式时，必须自己指定法线！！！否则没颜色
                surfaceTool.SetNormal <| origin + preDirection
                surfaceTool.AddVertex <| (origin + preDirection) * lonLatGrid.Radius
                surfaceTool.SetColor color
                surfaceTool.SetNormal <| origin + currentDirection
                surfaceTool.AddVertex <| (origin + currentDirection) * lonLatGrid.Radius

            preDirection <- currentDirection

    /// <summary>
    /// 绘制指定纬线
    /// </summary>
    /// <param name="env"></param>
    /// <param name="surfaceTool"></param>
    /// <param name="latitudeRadian">维度转为弧度制后的表示，+ 表示北纬，- 表示南纬（上方取正）</param>
    /// <param name="color"></param>
    /// <param name="dash">是否按虚线绘制</param>
    let private drawLatitude env (surfaceTool: SurfaceTool) (latitudeRadian: float32) (color: Color) (dash: bool) =
        let cos = Mathf.Cos latitudeRadian // 对应相比赤道应该缩小的比例
        let sin = Mathf.Sin latitudeRadian // 对应固定的高度
        // 本初子午线
        let primeMeridianDirection = Vector3(cos, 0f, 0f)
        // 西经 90 度
        let west90Direction = Vector3(0f, 0f, cos)

        draw90Degrees env surfaceTool color primeMeridianDirection west90Direction
        <| Vector3.Up * sin
        <| dash
        // 对向子午线
        let antiMeridianDirection = Vector3(-cos, 0f, 0f)

        draw90Degrees env surfaceTool color west90Direction antiMeridianDirection
        <| Vector3.Up * sin
        <| dash
        // 东经 90 度
        let east90Direction = Vector3(0f, 0f, -cos)

        draw90Degrees env surfaceTool color antiMeridianDirection east90Direction
        <| Vector3.Up * sin
        <| dash

        draw90Degrees env surfaceTool color east90Direction primeMeridianDirection
        <| Vector3.Up * sin
        <| dash

    /// <summary>
    /// 绘制指定经线
    /// </summary>
    /// <param name="env"></param>
    /// <param name="surfaceTool"></param>
    /// <param name="longitudeRadian">经度转为弧度制后的表示，+ 代表西经，- 代表东经（顺时针方向）</param>
    /// <param name="color"></param>
    let private drawLongitude env (surfaceTool: SurfaceTool) (longitudeRadian: float32) (color: Color) =
        let equatorDirection =
            Vector3(Mathf.Cos longitudeRadian, 0f, Mathf.Sin longitudeRadian)
        // 北面
        draw90Degrees env surfaceTool color equatorDirection Vector3.Up Vector3.Zero false
        // 南面
        draw90Degrees env surfaceTool color equatorDirection Vector3.Down Vector3.Zero false

    let doDraw (env: #ILonLatGridQuery) : DoDraw =
        fun () ->
            let lonLatGrid = env.LonLatGrid
            let surfaceTool = new SurfaceTool()
            surfaceTool.Begin Mesh.PrimitiveType.Lines

            for i in -180 + lonLatGrid.LongitudeInterval .. lonLatGrid.LongitudeInterval .. 180 do
                let longitudeRadian = Mathf.DegToRad(float32 i)

                let color =
                    if i = 0 || i = 180 then
                        lonLatGrid.MeridianColor
                    elif i = -90 || i = 90 then
                        lonLatGrid.Degree90LongitudeColor
                    elif i % (lonLatGrid.LongitudeInterval * lonLatGrid.DeeperLineInterval) = 0 then
                        lonLatGrid.DeeperLineColor
                    else
                        lonLatGrid.NormalLineColor

                drawLongitude env surfaceTool longitudeRadian color

            for i in -90 + lonLatGrid.LatitudeInterval .. lonLatGrid.LatitudeInterval .. 90 do
                let latitudeRadian = Mathf.DegToRad(float32 i)

                let color =
                    if i = 0 then
                        lonLatGrid.EquatorColor
                    elif i % (lonLatGrid.LatitudeInterval * lonLatGrid.DeeperLineInterval) = 0 then
                        lonLatGrid.CircleColor
                    else
                        lonLatGrid.NormalLineColor

                drawLatitude env surfaceTool latitudeRadian color false
            // 北极圈：北纬 66°34′
            if lonLatGrid.DrawArcticCircle then
                drawLatitude env surfaceTool
                <| Mathf.DegToRad 66.567f
                <| lonLatGrid.CircleColor
                <| true
            // 北回归线：北纬 23°26′
            if lonLatGrid.DrawTropicOfCancer then
                drawLatitude env surfaceTool
                <| Mathf.DegToRad 23.433f
                <| lonLatGrid.TropicColor
                <| true
            // 南回归线：南纬 23°26′
            if lonLatGrid.DrawTropicOfCapricorn then
                drawLatitude env surfaceTool
                <| Mathf.DegToRad -23.433f
                <| lonLatGrid.TropicColor
                <| true
            // 南极圈：南纬 66°34′
            if lonLatGrid.DrawAntarcticCircle then
                drawLatitude env surfaceTool
                <| Mathf.DegToRad -66.567f
                <| lonLatGrid.CircleColor
                <| true

            surfaceTool.SetMaterial lonLatGrid.LineMaterial
            lonLatGrid.MeshIns.Mesh <- surfaceTool.Commit()

    let drawOnPlanet (env: 'E when 'E :> IPlanetConfigQuery and 'E :> ILonLatGridQuery) : DrawOnPlanet =
        fun () ->
            let planetConfig = env.PlanetConfig
            env.LonLatGrid.Radius <- planetConfig.Radius + planetConfig.MaxHeight
            doDraw env ()

    let toggleFixFullVisibility (env: #ILonLatGridQuery) : ToggleFixFullVisibility =
        fun toggle -> env.LonLatGrid.FixFullVisibility <- toggle

    let onCameraMoved (env: #ILonLatGridQuery) : OnCameraMoved =
        fun _ delta ->
            let lonLatGrid = env.LonLatGrid

            if lonLatGrid.FixFullVisibility then
                () // 锁定显示时不处理事件
            else
                lonLatGrid.Show()
                lonLatGrid.Visibility <- lonLatGrid.Visibility + delta / lonLatGrid.FullVisibilityTime
                lonLatGrid.FadeVisibility <- false
                lonLatGrid.SetProcess true
